### ISSUE TITLE
[BUG] TCP connections broken

### ISSUE THREAD

**Issue by danpascu** [OpenSIPS/opensips#1663]
**OpenSIPS version you are running**

```
version: opensips 3.0.0-dev (x86_64/linux)
flags: STATS: On, SHM_EXTRA_STATS, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, QM_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 588bb6450
main.c compiled on  with gcc 8
```

**Describe the bug**

Have 2 proxies P1 and P2, with P1 forwarding to P2 using t_relay(). Open a TCP connection to P1, send a REGISTER request and read back the 401 reply then close the connection (in this example we'll use a REGISTER, but the problem manifests with any request type).

At this point any new TCP connection open from this client to P1 within the next 5 seconds will not get any answer from P1 when sending a new REGISTER (it'll just wait forever and nothing comes back), while on P1 I see a connection in the state SYN_SENT from P1 back to my address (using lsof). However a wireshark trace on P1 shows no TCP SYN packet being sent from P1.

The socket that shows up in SYN_SENT in lsof disappears (is closed) after 60-90 seconds.

After 5 seconds, a new connection works again, but it re-triggers the problem for another 5 seconds and so on.

This can be used as an attack vector by very quickly exhausting the available file descriptors and keeping them locked in this SYN_SENT state for very long times.

**To Reproduce**

See description above.

**Expected behavior**

I expect P1 to take the new TCP connections within 5 seconds and answer them and not attempt to make a connection back and be stuck in SYN_SENT for a minute or two.

**Relevant System Logs**

I do not see anything relevant in the logs.

**OS/environment information**
 - Operating System: `Debian unstable`
 - OpenSIPS installation: `manually compiled debian packages from git`
 - other relevant information:

relevant settings:

```
listen = udp:eth0:5060
listen = tcp:eth0:5060
listen = tls:eth0:5061

tos = TOS

tcp_accept_aliases  = no
tcp_connection_lifetime = 3600
tcp_keepalive = yes
tcp_keepcount = 4
tcp_keepidle = 60
tcp_keepinterval = 20

dns = yes
rev_dns = no
dns_use_search_list = no
dns_retr_time = 5
dns_retr_no = 1
dns_servers_no  = 1
disable_dns_blacklist = yes

modparam("proto_tcp", "tcp_async", 1)
modparam("proto_tcp", "tcp_max_msg_chunks", 50)
modparam("proto_tls", "tls_max_msg_chunks", 50)
```

**Additional context**

**Comment by razvancrainea** [comment]
Hello, Dan!

Any chance you could run `opensipsctl trap` during those 5 seconds, to see if any process is stuck doing something else?

Best regards,
Răzvan

**Comment by danpascu** [comment]
Here it is. I forgot to mention, the process that has the socket in SYN_SENT is TCP Main.

[gdb_opensips_20190410_104226.log](https://github.com/OpenSIPS/opensips/files/3062899/gdb_opensips_20190410_104226.log)

**Comment by bogdan-iancu** [comment]
Hi @danpascu,

I tried to reproduce this, but without luck :(...  I used an `sipp` script
[uac.xml.txt](https://github.com/OpenSIPS/opensips/files/3201834/uac.xml.txt) to send an OPTIONS via TCP.
Cmd line is `sipp -sf /tmp/uac.xml opensips_ip:5060 -t tn  -m 1`.

Also monitored the connections/sockets via `lsof -iTCP` and tcpdump (to check the proper closing of the TCP connection at TCP level).

But I haven't noticed the behavior you described.

Could you point to relevant details on how to do it ?

Also, do you have a tcpdump (IP level) for such a connection, just to check it ? (you can PM  to me)

**Comment by danpascu** [comment]
I didn't use sipp, I used a custom script to be able to set SO_LINGER properly to make it close TCP with a reset instead of a fin. Try the attached script (adjust the host variable to point to the hostname/ip of your proxy that only has a t_relay() in its main route). Later I learned that it doesn't seem to matter if I set SO_LINGER and close with a RST instead of FIN, it seems to behave all the same no matter how I close.


[regtest.txt](https://github.com/OpenSIPS/opensips/files/3202174/regtest.txt)

**Comment by bogdan-iancu** [comment]
I used your script while checking with both `netstat` and `lsof`. Still I fail to see the trailing socket.
Using:
`watch -n0.5 "netstat -atnp | grep opensips"`
and 
`watch -n0.5 "lsof -iTCP  | grep opensips "`

I can see the connection when the req/rpl happens, but nothing odd after that.

**Comment by bogdan-iancu** [comment]
How do you monitor for the socket ?

**Comment by danpascu** [comment]
Did the proxy you tested only do t_relay() in the main route towards another proxy? It only happens when 1 proxy uses t_relay to forward to another, not if the proxy processes the message itself.

Have another proxy that listens on tcp and only have this in the main route and not have any domains defined in the config or the domains table:

route {
  t_relay();
}

Then send a register with user@opensips.org over TCP with my script towards this proxy and it should forward it to your main proxy. Alternatively you can just use the script as it is, because it will forward to my sip2sip.info proxy. Just replace the host variable with your relaying proxy IP.

I used watch -n .1 'sudo lsof -np pid_of_tcp_main | grep TCP | tail -20'

**Comment by bogdan-iancu** [comment]
OK, did a proxy kind of setup and I got some progress here - I managed to reproduce an issue, but not sure if the right issue :D .

Re-running your script in less than 5 secs interval triggers an TCP error:

> ERROR:core:tcpconn_async_connect: poll error: flags 1c
> ERROR:core:tcpconn_async_connect: failed to retrieve SO_ERROR [server=_UAC_IP_:45566] (111) Connection refused
> ERROR:core:proto_tcp_send: async TCP connect failed
> ERROR:tm:msg_send: send() to _UAC_IP_:45566 for proto tcp/2 failed

What actually happens is: when re-sending the REGISTER too fast, the next attempt is seen by TM as a re-transmission (as your script does not change the VIA hdr).  And because it is a re-transmission, the `t_relay()` tries to automatically send back to UAC (python script) the last reply it sent back - and this is done without any further attempt of relaying.
But the transaction has a different src port stored for the REGISTER (it stores the src of the original request), so it tries to re-send back the reply (to UAC) to a wrong port (to the port of the TCP conn belonging to the original request, instead of the port of the TCP conn belonging to the re-transmission).
In my case, the original REGISTER is sent from `45566` and the re-transmission is sent from ` 45568`, so OpenSIPS tries to re-send back the reply to `45566`, but there is no conn anymore there and the connect also fails.
The 5 secs delay (for a successful re-attempt) is actually the WAIT time in TM (for deleting an inactive transaction). After that, any re-attempt is a new transaction, not a re-transmission.

Is this also your case ??

**Comment by danpascu** [comment]
It may very well be. I changed the script to randomize the Via on every call and it doesn't exhibit the behavior anymore. Unfortunately I cannot test it with the full configuration because I only have 1 proxy running the new version. In the full config I set the flag to not reopen TCP connections, so I'd like to see what it does with that. In the older versions, despite having that flag, it would still exhibit the problem, but in a worse form: there the connections would never be deleted, instead they would switch from "SYN sent" to a zombie after a minute or 2 and stay like that forever until the proxy ran out of file descriptors.

I'll check to see if setting the flag changes anything.
